Merge the manual and script plugins

bmw opened the pull request:
Fixes #2505, #3767 

This merges the manual and script plugins into one. While this was largely just refactoring existing code, I changed a few things about how the plugins used to work. I wanted to bring these things to the reviewer's attention to explain myself and to make sure they saw them.

#### --manual-test-mode

I have removed this flag entirely. It existed before to give us a way to test the manual plugin in integration tests. Now that the manual plugin accepts hooks, this is unnecessary. I did some searching to make sure that no one was using this flag and I couldn't find any references is to it other than output from `certbot --help` simply showing it exists. Because of this and the fact the flag was explicitly listed for testing, I think it is safe to simply remove this flag.

#### Stop using `simple_verify`

This fixes #2505. If you're not familiar, the `simple_verify` methods are client side checks that the challenge is set up correctly. I think doing this is bad for a number of reasons:

1. Code complexity for little/no benefit. Why does it matter if the client or the server tells you there is a problem? We don't do anything extra for the user if we detect a problem.
2. It's unreliable. Due to things like firewall rules, we've had `simple_verify` tell the user the challenge was set up incorrectly, yet it passes when Let's Encrypt checks. Here's [one example](https://github.com/certbot/certbot/issues/1777#issuecomment-162240317) of this.
3. For the DNS challenge, it's progress on #3650 because we're no longer using `dnspython` in Certbot.

#### Challenge selection for the script plugin

The plugin doesn't actually have to do anything here. The relevant code for this is [here](https://github.com/certbot/certbot/blob/master/certbot/auth_handler.py#L255). The challenges are preferred in the order given by the user with any challenges not supported by the selected plugin removed.

bmw commented:
Unfortunately I didn't get a chance to track down why the new integration tests are failing. I'll take a look next week.

bmw commented:
I couldn't get the DNS auth hook to work during integration tests. I created an issue to track it in #3901.

I don't believe that should be a blocker for this PR though. I've successfully tested the plugin with DNS challenges manually and with hooks.

pde commented on certbot/plugins/manual.py:
```diff
@@ -59,203 +52,113 @@ class Authenticator(common.Plugin):
+
+    def prepare(self):  # pylint: disable=missing-docstring
+        if self.config.noninteractive_mode and not self.conf('auth-hook'):
+            raise errors.PluginError(
```
Maybe use `errors.MissingCommandLineFlag`?  Or even make `NoninteractiveDisplay._interaction_fail` a public method, and call it?

bmw replied:
`prepare` has a very specific interface as defined in `interfaces.py` and code like [this](https://github.com/certbot/certbot/blob/master/certbot/plugins/disco.py#L114) only handles those specific error cases. I'd be open to raising `errors.MisconfigurationError` if you think it's more fitting though.

pde replied:
Would we break things by adding new exception types to the interface?

bmw replied:
No, but we'd need to document them in `interfaces.py` and add code to `plugins/disco.py` to handle the new cases. It is my understanding that the code was originally written this way to differentiate between cases where the plugin was simply unavailable and those where there was a real problem and Certbot should crash and notify the user.

Do you think making these changes is valuable? Should we be removing the feature of differentiating between exception types? Do you think these changes should be made in this PR?

My personal opinion is the message raised here is nicer than the one raised out of `_interaction_fail` because it's customized to this specific problem. I also care very little about the exception type, however, it's not immediately clear to me we should be ripping out the feature I described above just to use a different error type.

bmw commented on certbot/plugins/manual.py:
```diff
@@ -59,203 +52,113 @@ class Authenticator(common.Plugin):
-    def prepare(self):  # pylint: disable=missing-docstring,no-self-use
-        if self.config.noninteractive_mode and not self.conf("test-mode"):
-            raise errors.PluginError("Running manual mode non-interactively is not supported")
+        add('auth-hook',
```
According to your comment [here](https://github.com/certbot/certbot/issues/3767#issue-187879645):
> We disagreed on whether --auth-hook should always be run or whether it should be specific to --manual, but concluded that this would be the prerogative of whoever implements this refactor.

Both Joona and I argued against doing this in #3767.

pde replied:
OK I do see meeting minutes noting that it was up to whoever implemented this to choose between `--auth-hook` and `--manual-auth-hook` :)

bmw commented on certbot/plugins/manual.py:
```diff
@@ -59,203 +52,113 @@ class Authenticator(common.Plugin):
+    def _verify_ip_logging_ok(self):
+        if not self.conf('public-ip-logging-ok'):
+            cli_flag = '--{0}'.format(self.option_name('public-ip-logging-ok'))
+            msg = ('NOTE: The IP of this machine will be publicly logged as '
```
I did this because I prefer single quotes for strings, however, the lines using double quotes include single quote characters. Rather than escaping them, I just used double quotes which is consistent with [our](https://certbot.eff.org/docs/contributing.html#coding-style) [style](https://www.python.org/dev/peps/pep-0008/#string-quotes) [guides](https://google.github.io/styleguide/pyguide.html?showone=Strings#Strings).

joohoi replied:
This is ok, though not my personal preference. Have to appreciate underlining the legibility with three different sources ;)

bmw commented:
This is ready for another review. I've fixed or replied to all review comments.

In fixing the incorrect error message for hooks, I was able to delete a lot of code. The one additional behavioral change I made here to simplify the implementation is we no longer include the CLI flag with the invalid command in the raised error message. I don't think this really hurts usability but if the reviewer does, I'll either add code to catch and reraise the error in `validate_hooks` or write an [argparse.Action](https://docs.python.org/2.7/library/argparse.html#argparse.Action) class in `util` that all hooks can use so we can stop passing around hook prefixes to be included in potential error messages (the `argparse.Action` object already has access to the CLI flag that was used).

pde commented on certbot/plugins/manual.py:
```diff
@@ -59,203 +52,111 @@ class Authenticator(common.Plugin):
-            raise errors.PluginError("Running manual mode non-interactively is not supported")
+        add('auth-hook',
+            help='Path or command to execute for the authentication script')
+        add('cleanup-hook',
```
If this is landing after #3918, please add some checks for these hook to [the new hook integration tests](https://github.com/certbot/certbot/pull/3918/files#diff-c74d3ad4d86d0804f64e9354fa460b94R55).

bmw replied:
Ugh. I just realized my hook error message improvements and refactoring conflicts pretty badly with that PR.

bmw replied:
Also, do you think this is necessary? I already wrote [integration tests for the manual hooks](https://github.com/certbot/certbot/pull/3890/files#diff-c74d3ad4d86d0804f64e9354fa460b94R47) that are actually run to perform the challenge.

bmw commented:
I went ahead and removed the hook improvement that I implemented. I think making these changes is outside of the scope of this PR and I don't want the merge conflicts it creates with #3918 to prevent this from being merged when it is one of the last large PRs for 0.10.0. I created #3939 to track hook error messages as a separate issue.

I also added another commit to make the name of the hook used in the error messages thrown in `validate_hooks` consistent with other hooks.

joohoi commented:
LGTM!
